File: backend/quiz-service/src/main/java/io/github/krivolapovdev/codeoutputquiz/quizservice/service/QuizMetaService.java
package io.github.krivolapovdev.codeoutputquiz.quizservice.service;

import io.github.krivolapovdev.codeoutputquiz.quizservice.enums.DifficultyLevel;
import io.github.krivolapovdev.codeoutputquiz.quizservice.enums.ProgrammingLanguage;
import io.github.krivolapovdev.codeoutputquiz.quizservice.mapper.ProgrammingLanguageMapper;
import io.github.krivolapovdev.codeoutputquiz.quizservice.response.ProgrammingLanguageResponse;
import java.util.List;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
@Slf4j
@RequiredArgsConstructor
public class QuizMetaService {
  private final ProgrammingLanguageMapper programmingLanguageMapper;

  public Mono<List<ProgrammingLanguageResponse>> getSupportedProgrammingLanguages() {
    log.info("Getting supported programming languages");
    return Flux.fromArray(ProgrammingLanguage.values())
        .map(programmingLanguageMapper::toResponse)
        .collectList();
  }

  public Mono<List<DifficultyLevel>> getSupportedDifficultyLevels() {
    log.info("Getting supported difficulty levels");
    return Flux.fromArray(DifficultyLevel.values()).collectList();
  }
}

File: frontend/src/shared/api/quiz/quizMetaService.ts
import { api } from "../base";
import type { ProgrammingLanguageResponse } from "./types";

const baseURL = "/api/v1/quizzes/meta";

export const quizMetaService = {
  fetchSupportedProgrammingLanguages: async (): Promise<
    ProgrammingLanguageResponse[]
  > => {
    const { data } = await api.get<ProgrammingLanguageResponse[]>(
      `${baseURL}/supported-programming-languages`
    );
    return data;
  },

  fetchSupportedDifficultyLevels: async (): Promise<string[]> => {
    const { data } = await api.get<string[]>(
      `${baseURL}/supported-difficulty-levels`
    );
    return data;
  }
};
